type.go

```diff
@@ -0,0 +1,458 @@
+	if t.Kind() != reflect.Ptr || t.Elem().Kind() != reflect.Struct {
+		return nil, fmt.Errorf("type is not pointer to struct")
+	}
+	hasBody := false
+	var pt requestType
+	for _, f := range fields(t.Elem()) {
+		tag, err := parseTag(f.Tag, f.Name)
+		if err != nil {
+			return nil, errgo.Notef(err, "bad tag %q in field %s", f.Tag, f.Name)
+		}
+		if tag.source == sourceBody {
+			if hasBody {
+				return nil, errgo.New("more than one body field specified")
+			}
+			hasBody = true
+		}
+		field := field{
+			index: f.Index,
+		}
+		if f.Type.Kind() == reflect.Ptr {
+			// The field is a pointer, so when the value is set,
+			// we need to create a new pointer to put
+			// it into.
+			field.makeResult = makePointerResult
+			f.Type = f.Type.Elem()
+		} else {
+			field.makeResult = makeValueResult
+		}
+		field.unmarshal, err = getUnmarshaler(tag, f.Type)
+		if err != nil {
+			return nil, errgo.Mask(err)
+		}
+		if f.Anonymous {
+			if tag.source != sourceBody && tag.source != sourceNone {
+				return nil, errgo.New("httprequest tag not yet supported on anonymous fields")
+			}
+		}
+		pt.fields = append(pt.fields, field)
+	}
+	return &pt, nil
+}
+
+func makePointerResult(v reflect.Value) reflect.Value {
+	if v.IsNil() {
+		v.Set(reflect.New(v.Type().Elem()))
+	}
+	return v.Elem()
+}
+
+func makeValueResult(v reflect.Value) reflect.Value {
+	return v
+}
+
+// getUnmarshaler returns an unmarshaler function
+// suitable for unmarshaling a field with the given tag
+// into a value of the given type.
+func getUnmarshaler(tag tag, t reflect.Type) (unmarshaler, error) {
+	switch {
+	case tag.source == sourceNone:
+		return unmarshalNop, nil
+	case tag.source == sourceBody:
+		return unmarshalBody, nil
+	case t == reflect.TypeOf([]string(nil)):
+		if tag.source != sourceForm {
+			return nil, errgo.New("invalid target type []string for url parameter")
+		}
+		return unmarshalAllField(tag.name), nil
+	case t == stringType:
+		return unmarshalString(tag), nil
+	case implementsTextUnmarshaler(t):
+		return unmarshalWithUnmarshalText(t, tag), nil
+	default:
+		return unmarshalWithScan(tag), nil
+	}
+}
+
+// unmarshalNop just creates the result value but does not
+// fill it out with anything. This is used to create pointers
+// to new anonymous field members.
+func unmarshalNop(v reflect.Value, p Params, makeResult resultMaker) error {
+	makeResult(v)
+	return nil
+}
+
+// unmarshalAllField unmarshals all the form fields for a given
+// attribute into a []string slice.
+func unmarshalAllField(name string) unmarshaler {
+	return func(v reflect.Value, p Params, makeResult resultMaker) error {
+		vals := p.Form[name]
+		if len(vals) > 0 {
+			makeResult(v).Set(reflect.ValueOf(vals))
+		}
+		return nil
+	}
+}
+
+// unmarshalString unmarshals into a string field.
+func unmarshalString(tag tag) unmarshaler {
+	getVal := formGetters[tag.source]
+	if getVal == nil {
+		panic("unexpected source for unmarshal text field")
```

Comment:
copy and paste error? is this really the right error message perhaps "unexpected source for string field"?